### PR TITLE
Added bls aggregate sig check for block validation

### LINKED ISSUE #215: Retrieve BLS Public Key
**Task summary**

Need the ability to retrieve the public key from a BLS signature, this functionality is needed to verify the signature of messages. 

Required for the check block messages part of block validation, specifically for bls messages as the public keys are a required param for the [`verify_bls_aggregate`](https://github.com/ChainSafe/forest/blob/master/crypto/src/signature.rs#L155) function. 

**Other information and links**

- [Lotus GetBlsPublicKey](https://github.com/filecoin-project/lotus/blob/master/chain/stmgr/stmgr.go#L369:25)
- [Lotus GetBlsPublicKey usage](https://github.com/filecoin-project/lotus/blob/testnet/3/chain/sync.go#L816)

In our codebase, this would be used during block validation process in the `Chainsyncer` component [here](https://github.com/ChainSafe/forest/blob/master/blockchain/chain_sync/src/sync.rs#L489).

### PR DESCRIPTION
**Pull request by dutterbutter**
**Summary of changes**
Changes introduced in this pull request:
- Added `get_bls_public_key` method 
- Added the block message check verifying bls aggregate 

**Reference issue to close (if applicable)**

Closes #215 

**Other information and links**

- [Block Message Check reference](https://github.com/filecoin-project/lotus/blob/testnet/3/chain/sync.go#L778)

### CODE REVIEW

File: blockchain/chain_sync/src/sync.rs [hunk 1]
```diff
@@ -514,12 +516,34 @@ where
...
+        for m in block.bls_msgs() {
+            let pk = self
+                .state_manager
+                .get_bls_public_key(m.from(), tip.parent_state())?;
+            pub_keys.push(pk);
+            cids.push(m.cid()?.to_bytes());
+        }
+        if let Some(sig) = block.header().bls_aggregate() {
+            if !verify_bls_aggregate(
+                cids.iter()
```
Comment (ec2) [inline blockchain/chain_sync/src/sync.rs:531]: @austinabell If you can find a better way to get a Vec<Vec<u8>> to a &[&[u8]] I'll give you $5

Comment (austinabell) [reply blockchain/chain_sync/src/sync.rs:531]: nope, that's the best way possible I think, unfortunate there isn't a cleaner way

File: blockchain/chain_sync/src/sync.rs [hunk 2]
```diff
@@ -514,12 +516,34 @@ where
...
+        let mut pub_keys = Vec::new();
+        let mut cids = Vec::new();
+        for m in block.bls_msgs() {
+            let pk = self
+                .state_manager
+                .get_bls_public_key(m.from(), tip.parent_state())?;
+            pub_keys.push(pk);
+            cids.push(m.cid()?.to_bytes());
+        }
+        if let Some(sig) = block.header().bls_aggregate() {
```
Comment (austinabell) [inline blockchain/chain_sync/src/sync.rs:529]: Should probably error if there is no bls aggregate signature on the block, right?

File: blockchain/state_manager/src/lib.rs [hunk 3]
```diff
@@ -83,4 +83,17 @@ where
             ))
         }
     }
+    /// Returns a bls public key from provided address
+    pub fn get_bls_public_key(&self, addr: &Address, state_cid: &Cid) -> Result<Vec<u8>, Error> {
```
Comment (austinabell) [inline blockchain/state_manager/src/lib.rs:87]: Just a note that the public key should probably be an array (I'll make a pass through the Address crate soon I guess)

**Pull request merged**